### PR TITLE
Soundness: `Robj` must be transparent

### PR DESCRIPTION
**Pull request by CGMossa**
It is possible to use `std::mem::transmute` to go from a `SEXP` to, say, `&Robj`/`&mut Robj`. This is indeed necessary to make some API possible. See `Index` on `List` as an example. A follow-up PR for `Index` on `List` will come..

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +19,9 @@ return `Option<Strings>` instead of opaque `Robj`.
 
 ### Deprecated
 
+- Removed `from_sexp_ref` from public API. This is a completely unsound function, and uses of it must not
```
Comment (Ilia-Kosenkov) [inline CHANGELOG.md:22]: Can you rephrase this please? I think something is not right in this sentence.

Comment (JosiahParry) [reply CHANGELOG.md:22]: What about: 

```
- Removed `from_sexp_ref` from the public API. `from_sexp_ref()` is an unsafe method and is intended for use only within `extendr-api`.

Comment (JosiahParry) [inline CHANGELOG.md:22]:
```suggestion
- Removed `from_sexp_ref()` from the public API. `from_sexp_ref()` is an unsafe method and is for internal use only.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -20,6 +19,9 @@ return `Option<Strings>` instead of opaque `Robj`.
...
+used otherwise. [[#855]](https://github.com/extendr/extendr/pull/855)
```
Comment (JosiahParry) [inline CHANGELOG.md:23]:
```suggestion
```

**Review by JosiahParry — APPROVED** [review]
Tests that fail are on r-devel due to BDR's new check-rust function. We have an existing patch but it is being ignored. https://bugs.r-project.org/show_bug.cgi?id=18806

This is good to merge now. Thanks for the help everyone
